### ISSUE TITLE
Dubious implementation of ROIAlign in 'max' mode

### ISSUE THREAD

**Issue by yury-intel** [microsoft/onnxruntime#6146]
There are cases, when the results of ROIAlign may be significally different from expected.

Implementation of ROIAlign runs questionably in MAX mode.
I.e., when weighed data feature_map is collected, we commit MAX - operation to all collected data.
So, we return one of weighed value (by [0, 1] - coefficient), which is assuredly less, then all interpolated values.
In the case of centered coordinates of sampling point, nearly 1/4 from expected value is returned.
We need instead, at first, commit interpolation, and afterwards given type of operation (AVG or MAX).
Suggested algorithm is similar to Caffe implementation.

![image](https://user-images.githubusercontent.com/57213286/102330641-dc643400-3f9a-11eb-961f-2dfe4cba37cc.png)
![image](https://user-images.githubusercontent.com/57213286/102331251-a70c1600-3f9b-11eb-9390-72cf6291b989.png)

**Comment by pranavsharma** [comment]
Would you like to contribute?

**Comment by fdwr** [comment]
@yuri-intel: I found some alignment issues too (comparing against PyTorch, the Mask RCNN paper, and FB research's Detectron 2). Might be the same cause, a half pixel off? https://github.com/microsoft/onnxruntime/issues/6921

**Comment by hariharans29** [comment]
Not quite following this. Can you quantify "running questionably" with a simple test case ? For a given input, what do you expect to see vs what actually gets produced ?

**Comment by fdwr** [comment]
Yury: I found a bug in `max` mode after running our WindowsAI ONNX conformance tests, which _might_ fix your issue too:
https://github.com/microsoft/onnxruntime/pull/7354

**Comment by masahi** [comment]
Yes I remember this was a bug in ORT `roi_align` with max mode, where summation (for averaging) for bilinear interpolation is incorrrectly replaced with max.

**Comment by CBlackBird** [comment]
> Yes I remember this was a bug in ORT `roi_align` with max mode, where summation (for averaging) for bilinear interpolation is incorrrectly replaced with max. Maximization should be done on interpolated values, interpolation itself is always sum (averaging).

i found same issue, does onnxruntime will fix it on future?

**Comment by fdwr** [comment]
@[hariharans29](https://github.com/hariharans29) I'm unsure who the "owner" of CPU RoiAlign is, but I'll try to push my old PR again from above. Last time some tests failed because they were based on the bad expected results using the old numbers, and I'm unsure how they were generated in the first place. So I'm unsure what to do there, besides disabling those bad tests 🤔. I'll sync with master to see what new/old failures surface.
